python/ray/data/datasource/file_based_datasource.py

```diff
@@ -556,7 +556,18 @@ def _resolve_paths_and_filesystem(
             filesystem = resolved_filesystem
         elif need_unwrap_path_protocol:
             resolved_path = _unwrap_protocol(resolved_path)
-        resolved_path = filesystem.normalize_path(resolved_path)
+
+        # S3 bucket name cannot contain "@", thus indicating pyarrow failed
+        # to parse a path with ananymous credentials, for a path like
+        # `s3://anonymous@bucket/data.parquet` with known filesystem, pyarrow
+        # returns `anonymous@bucket/data.parquet` as the resolved path by
+        # mistake.
+        if type(filesystem).__name__ == "S3FileSystem" and "@" in resolved_path:
```

Comment:
Any reason to not do the more Pythonic instance check here?
```suggestion
        if isinstance(filesystem, S3FileSystem) and "@" in resolved_path:
```

Reply:
@clarkzinzow so i looked into pyarrow's implementation a bit and it only lazily imports `S3FileSystem` and there's no where that assumes it's always available, so i fell back to use string check. If dataset can safely assume we always have this module i can change it back to instance check

Reply:
Ah yeah the `S3FileSystem` does some expensive things (e.g. network calls) at initialization time, so the string check would probably be best. Thank you for looking into that!

Reply:
Pyarrow's python implementation ... makes me appreciate ray dataset code a lot more now lol